syntax = "proto3";
package tinykvpb;

import "kvrpcpb.proto";
import "raft_serverpb.proto";
import "raft_cmdpb.proto";
import "coprocessor.proto";

import "gogoproto/gogo.proto";

option (gogoproto.sizer_all) = true;
option (gogoproto.marshaler_all) = true;
option (gogoproto.unmarshaler_all) = true;

// Serve as a distributed kv database. See the request and response definitions in
// kvrpcpb.proto
service TinyKv {
    // KV commands with mvcc/txn supported.
    rpc KvGet(kvrpcpb.GetRequest) returns (kvrpcpb.GetResponse) {}
    rpc KvScan(kvrpcpb.ScanRequest) returns (kvrpcpb.ScanResponse) {}
    rpc KvPrewrite(kvrpcpb.PrewriteRequest) returns (kvrpcpb.PrewriteResponse) {}
    rpc KvCommit(kvrpcpb.CommitRequest) returns (kvrpcpb.CommitResponse) {}
    rpc KvCheckTxnStatus(kvrpcpb.CheckTxnStatusRequest) returns (kvrpcpb.CheckTxnStatusResponse) {}
    rpc KvBatchRollback(kvrpcpb.BatchRollbackRequest) returns (kvrpcpb.BatchRollbackResponse) {}
    rpc KvResolveLock(kvrpcpb.ResolveLockRequest) returns (kvrpcpb.ResolveLockResponse) {}

    // transfer leader request
    rpc TransferLeader(raft_cmdpb.TransferLeaderRequest) returns (raft_cmdpb.TransferLeaderResponse) {}

    // conf change request
    rpc PeerConfChange(raft_cmdpb.ChangePeerRequest) returns (raft_cmdpb.ChangePeerResponse) {}
    
    // split segion request
    rpc SplitRegion(raft_cmdpb.SplitRequest) returns (raft_cmdpb.SplitResponse) {}
    
    // RawKV commands.
    rpc RawGet(kvrpcpb.RawGetRequest) returns (kvrpcpb.RawGetResponse) {}
    rpc RawPut(kvrpcpb.RawPutRequest) returns (kvrpcpb.RawPutResponse) {}
    rpc RawDelete(kvrpcpb.RawDeleteRequest) returns (kvrpcpb.RawDeleteResponse) {}
    rpc RawScan(kvrpcpb.RawScanRequest) returns (kvrpcpb.RawScanResponse) {}

    // Raft commands (tinykv <-> tinykv).
    rpc Raft(raft_serverpb.RaftMessage) returns (raft_serverpb.Done) {}
    rpc Snapshot(raft_serverpb.SnapshotChunk) returns (raft_serverpb.Done) {}
    
    // Coprocessor 
    rpc Coprocessor(coprocessor.Request) returns (coprocessor.Response) {}
}
